cmd/disk-cache.go

```diff
@@ -0,0 +1,490 @@
+	dir string // caching directory (disk path in the --cache-policy file)
+	// Max disk usage limit (--cache-max) in percent. If the disk usage crosses this
+	// limit then we stop caching.
+	maxUsage int
+
+	// creation-time of the cache
+	createTime time.Time
+	// cache object
+	db ObjectLayer
+
+	// purge() listens on this channel to start the cache-purge process
+	purgeChan chan struct{}
+}
+
+// cache policy for prefix
+type cachePolicy struct {
+	cachePrefix string
+}
+
+// mapping of disk path to diskCache object
+type diskToCacheMap map[string]*diskCache
+
+// mapping of cache policy to disk path
+type policyToDiskMap map[cachePolicy]string
+
+// Returns if the disk usage is low.
+// Disk usage is low if usage is < 80% of maxUsage
+// Ex. for a 100GB disk, if maxUsage is configured as 70% then maxUsage is 70G
+// hence disk usage is low if the disk usage is less than 56G (because 80% of 70G is 56G)
+func (c diskCache) diskUsageLow() bool {
+	minUsage := c.maxUsage * 80 / 100
+
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		errorIf(err, "Error getting disk information on %s", c.dir)
+		return false
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) < minUsage
+}
+
+// Return if the disk usage is high.
+// Disk usage is high if disk used is > maxUsage
+func (c diskCache) diskUsageHigh() bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		errorIf(err, "Error getting disk information on %s", c.dir)
+		return true
+	}
+	usedPercent := (di.Total - di.Free) * 100 / di.Total
+	return int(usedPercent) > c.maxUsage
+}
+
+// Returns if size space can be allocated without exceeding
+// max disk usage
+func (c diskCache) diskAvailable(size int64) bool {
+	di, err := disk.GetInfo(c.dir)
+	if err != nil {
+		errorIf(err, "Error getting disk information on %s", c.dir)
+		return false
+	}
+	usedPercent := (di.Total - di.Free - uint64(size)) * 100 / di.Total
+	return int(usedPercent) < c.maxUsage
+}
+
+// Purge cache entries that were not accessed.
+func (c diskCache) purge() {
+	delimiter := "/"
+	maxKeys := 1000
+	for {
+		for {
+			if c.diskUsageLow() {
+				break
+			}
+			// expire objects older than one month since accessed
+			expiry := time.Now().AddDate(0, 0, -30)
+			deletedCount := 0
+			buckets, err := c.db.ListBuckets()
+			if err != nil {
+				errorIf(err, "Unable to list buckets.")
+			}
+			for _, bucket := range buckets {
+				var continuationToken string
+				var marker string
+				for {
+					objects, err := c.db.ListObjects(bucket.Name, marker, continuationToken, delimiter, maxKeys)
+					if err != nil {
+						break
+					}
+
+					if !objects.IsTruncated {
+						break
+					}
+					marker = objects.NextMarker
+					for _, object := range objects.Objects {
+						if object.AccTime.After(expiry) {
+							continue
+						}
+						if err = c.db.DeleteObject(bucket.Name, object.Name); err != nil {
+							errorIf(err, "Unable to remove cache entry in dir %s/%s", bucket.Name, object.Name)
```

Comment:
Even if there is an error deletedCount is calculating ++ we can avoid that?

Reply:
fixed this.